File: Data Structures & Algorithms/target-sum/submission-2.cpp
class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target) {
        int n = nums.size();
        int total = accumulate(nums.begin(),nums.end(),0);
        vector<int> prev(total+1,0);
        prev[0] = 1;
        for(int i = 0;i<n;i++){
            vector<int> curr(total+1,0);
            for(int sm = 0;sm<=total;sm++){
                curr[sm] = prev[sm];
                if(sm >= nums[i]){
                    curr[sm] += prev[sm-nums[i]];
                }
            }
            prev = curr;
        }
        int ans = 0;
        // for(int i = 0;i<=total;i++){
        //     cout<<prev[i]<<" ";
        // }
        // cout<<endl;
        for(int sm = 0;sm<=total;sm++){
            int rest = total-sm;
            if(sm-rest == target){
                ans += prev[sm];
            }
        }
        return ans;
    }
};

File: Data Structures & Algorithms/search-2d-matrix/submission-2.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int n = matrix.length;
        int m = matrix[0].length;

        int st = 0;
        int end = m*n - 1;
        while(st <= end){
            int mid = (st+end)/2;
            int r = mid/m;
            int c = mid%m;
            if(matrix[r][c] == target){
                return true;
            }else if(matrix[r][c] > target){
                end = mid-1;
            }else{
                st = mid+1;
            }
        }
        return false;
    }
}
